cordapp/src/main/kotlin/com/r3/corda/finance/obligation/client/DummyContract.kt

```diff
@@ -0,0 +1,8 @@
+package com.r3.corda.finance.obligation.client
+
+import net.corda.core.contracts.Contract
+import net.corda.core.transactions.LedgerTransaction
+
```

Comment:
Why do we need this?

Reply:
Yah, I'm not quite sure how the jar attachment and class loader works here. I was getting errors about not finding the attachment for certain classes. I can try again without this.

Reply:
Tried again without the DummyContract and got:

```
[ERROR] 2019-05-24T17:45:40,068Z [Node thread-1] transactions.TransactionBuilder.addMissingAttachment - The transaction currently built is missing an attachment for class: com/r3/corda/finance/obligation/client/flows/MakeOffLedgerPayment.
                        Attempted to find a suitable attachment but could not find any in the storage.
                        Please contact the developer of the CorDapp for further instructions. {actor_id=admin, actor_owning_identity=CN=party2, O=Party2, L=Calgary, C=CA, actor_store_id=NODE_CONFIG, fiber-id=10000011, flow-id=659ad3e0-782d-499d-bf94-bcf82c121f1d, invocation_id=c7976465-392d-4ef9-8186-c114975b11a1, invocation_timestamp=2019-05-24T17:45:37.310Z, origin=admin, session_id=e2027977-e1bc-4464-af6a-05547df67cd8, session_timestamp=2019-05-24T17:44:14.174Z, thread-id=474}
[INFO ] 2019-05-24T17:45:40,068Z [Node thread-1] corda.flow.run - Flow raised an error: com/r3/corda/finance/obligation/client/flows/MakeOffLedgerPayment. Sending it to flow hospital to be triaged. {actor_id=admin, actor_owning_identity=CN=party2, O=Party2, L=Calgary, C=CA, actor_store_id=NODE_CONFIG, fiber-id=10000011, flow-id=659ad3e0-782d-499d-bf94-bcf82c121f1d, invocation_id=c7976465-392d-4ef9-8186-c114975b11a1, invocation_timestamp=2019-05-24T17:45:37.310Z, origin=admin, session_id=e2027977-e1bc-4464-af6a-05547df67cd8, session_timestamp=2019-05-24T17:44:14.174Z, thread-id=474}
```

and

```
java.lang.NoClassDefFoundError: com/r3/corda/finance/obligation/client/flows/MakeOffLedgerPayment
        at java.lang.ClassLoader.defineClass1(Native Method) ~[?:1.8.0_212]
        at java.lang.ClassLoader.defineClass(ClassLoader.java:763) ~[?:1.8.0_212]
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142) ~[?:1.8.0_212]
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:468) ~[?:1.8.0_212]
        at java.net.URLClassLoader.access$100(URLClassLoader.java:74) ~[?:1.8.0_212]
        at java.net.URLClassLoader$1.run(URLClassLoader.java:369) ~[?:1.8.0_212]
        at java.net.URLClassLoader$1.run(URLClassLoader.java:363) ~[?:1.8.0_212]
        at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_212]
        at java.net.URLClassLoader.findClass(URLClassLoader.java:362) ~[?:1.8.0_212]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_212]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:411) ~[?:1.8.0_212]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_212]
        at java.lang.Class.forName0(Native Method) ~[?:1.8.0_212]
        at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_212]
        at net.corda.serialization.internal.model.TypeIdentifier$Parameterised.getLocalType(TypeIdentifier.kt:209) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.model.TypeIdentifier$Parameterised.getLocalType(TypeIdentifier.kt:218) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.model.ClassCarpentingTypeLoader$load$noCarpentryRequired$1$1.apply(TypeLoader.kt:38) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.model.ClassCarpentingTypeLoader$load$noCarpentryRequired$1$1.apply(TypeLoader.kt:25) ~[corda-serialization-4.1-RC03.jar:?]
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660) ~[?:1.8.0_212]
        at net.corda.serialization.internal.model.ClassCarpentingTypeLoader$load$noCarpentryRequired$1.invoke(TypeLoader.kt:38) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.model.ClassCarpentingTypeLoader$load$noCarpentryRequired$1.invoke(TypeLoader.kt:25) ~[corda-serialization-4.1-RC03.jar:?]
        at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:149) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:109) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at kotlin.collections.MapsKt__MapsKt.putAll(Maps.kt:339) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at kotlin.collections.MapsKt__MapsKt.toMap(Maps.kt:504) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at kotlin.collections.MapsKt__MapsKt.toMap(Maps.kt:498) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at net.corda.serialization.internal.model.ClassCarpentingTypeLoader.load(TypeLoader.kt:45) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory.reflect(RemoteSerializerFactory.kt:130) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory.access$reflect(RemoteSerializerFactory.kt:48) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory$get$1.invoke(RemoteSerializerFactory.kt:72) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory$get$1.invoke(RemoteSerializerFactory.kt:48) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DefaultDescriptorBasedSerializerRegistry.getOrBuild(DescriptorBasedSerializerRegistry.kt:28) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory.get(RemoteSerializerFactory.kt:67) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposedSerializerFactory.get(SerializerFactory.kt) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.readObject$serialization(DeserializationInput.kt:172) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.readObjectOrNull$serialization(DeserializationInput.kt:147) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DescribedTypeReadStrategy.readProperty(ComposableTypePropertySerializer.kt:202) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableTypePropertySerializer.readProperty(ComposableTypePropertySerializer.kt) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableObjectReader$readObject$$inlined$ifThrowsAppend$lambda$1.invoke(ObjectSerializer.kt:140) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableObjectReader$readObject$$inlined$ifThrowsAppend$lambda$1.invoke(ObjectSerializer.kt:122) ~[corda-serialization-4.1-RC03.jar:?]
        at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:149) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at net.corda.serialization.internal.amqp.ComposableObjectReader.readObject(ObjectSerializer.kt:219) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableObjectSerializer.readObject(ObjectSerializer.kt:91) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.readObject$serialization(DeserializationInput.kt:182) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.readObjectOrNull$serialization(DeserializationInput.kt:147) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DescribedTypeReadStrategy.readProperty(ComposableTypePropertySerializer.kt:202) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableTypePropertySerializer.readProperty(ComposableTypePropertySerializer.kt) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableObjectReader$readObject$$inlined$ifThrowsAppend$lambda$1.invoke(ObjectSerializer.kt:140) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableObjectReader$readObject$$inlined$ifThrowsAppend$lambda$1.invoke(ObjectSerializer.kt:122) ~[corda-serialization-4.1-RC03.jar:?]
        at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:149) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at net.corda.serialization.internal.amqp.ComposableObjectReader.readObject(ObjectSerializer.kt:219) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.ComposableObjectSerializer.readObject(ObjectSerializer.kt:91) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.readObject$serialization(DeserializationInput.kt:182) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.readObjectOrNull$serialization(DeserializationInput.kt:147) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput$deserialize$1.invoke(DeserializationInput.kt:124) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.des(DeserializationInput.kt:99) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.DeserializationInput.deserialize(DeserializationInput.kt:119) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.amqp.AbstractAMQPSerializationScheme.deserialize(AMQPSerializationScheme.kt:225) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.SerializationFactoryImpl$deserialize$1$1.invoke(SerializationScheme.kt:105) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.core.serialization.SerializationFactory.withCurrentContext(SerializationAPI.kt:71) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.serialization.internal.SerializationFactoryImpl$deserialize$1.invoke(SerializationScheme.kt:105) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.serialization.internal.SerializationFactoryImpl$deserialize$1.invoke(SerializationScheme.kt:73) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.core.serialization.SerializationFactory.asCurrent(SerializationAPI.kt:85) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.serialization.internal.SerializationFactoryImpl.deserialize(SerializationScheme.kt:105) ~[corda-serialization-4.1-RC03.jar:?]
        at net.corda.core.internal.TransactionUtilsKt$deserialiseComponentGroup$1.invoke(TransactionUtils.kt:78) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.internal.TransactionUtilsKt$deserialiseComponentGroup$1.invoke(TransactionUtils.kt) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.internal.LazyMappedList.get(InternalUtils.kt:552) ~[corda-core-4.1-RC03.jar:?]
        at java.util.AbstractList$Itr.next(AbstractList.java:358) ~[?:1.8.0_212]
        at java.util.AbstractCollection.toArray(AbstractCollection.java:141) ~[?:1.8.0_212]
        at java.util.ArrayList.addAll(ArrayList.java:581) ~[?:1.8.0_212]
        at kotlin.collections.CollectionsKt___CollectionsKt.plus(_Collections.kt:1970) ~[kotlin-stdlib-1.2.71.jar:1.2.71-release-64 (1.2.71)]
        at net.corda.core.internal.Verifier.<init>(TransactionVerifierServiceInternal.kt:33) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.LedgerTransaction$internalPrepareVerify$1.invoke(LedgerTransaction.kt:148) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.LedgerTransaction$internalPrepareVerify$1.invoke(LedgerTransaction.kt:45) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder$withAttachmentsClassloaderContext$1.invoke(AttachmentsClassLoader.kt:339) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.serialization.SerializationFactory.withCurrentContext(SerializationAPI.kt:71) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder.withAttachmentsClassloaderContext(AttachmentsClassLoader.kt:338) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.serialization.internal.AttachmentsClassLoaderBuilder.withAttachmentsClassloaderContext$default(AttachmentsClassLoader.kt:314) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.LedgerTransaction.internalPrepareVerify$core(LedgerTransaction.kt:144) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.LedgerTransaction.verify(LedgerTransaction.kt:134) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.addMissingDependency(TransactionBuilder.kt:186) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core(TransactionBuilder.kt:166) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core(TransactionBuilder.kt:169) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core(TransactionBuilder.kt:169) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core(TransactionBuilder.kt:169) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.toWireTransactionWithContext$core$default(TransactionBuilder.kt:134) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.toWireTransaction(TransactionBuilder.kt:131) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.transactions.TransactionBuilder.toSignedTransaction(TransactionBuilder.kt:711) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.node.ServiceHub$DefaultImpls.signInitialTransaction(ServiceHub.kt:232) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.core.node.ServiceHub$DefaultImpls.signInitialTransaction(ServiceHub.kt:246) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.node.services.api.ServiceHubInternal$DefaultImpls.signInitialTransaction(ServiceHubInternal.kt) ~[corda-node-4.1-RC03.jar:?]
        at net.corda.node.internal.AbstractNode$ServiceHubInternalImpl.signInitialTransaction(AbstractNode.kt:962) ~[corda-node-4.1-RC03.jar:?]
        at net.corda.core.node.ServiceHub$DefaultImpls.signInitialTransaction(ServiceHub.kt:268) ~[corda-core-4.1-RC03.jar:?]
        at net.corda.node.services.api.ServiceHubInternal$DefaultImpls.signInitialTransaction(ServiceHubInternal.kt) ~[corda-node-4.1-RC03.jar:?]
        at net.corda.node.internal.AbstractNode$ServiceHubInternalImpl.signInitialTransaction(AbstractNode.kt:962) ~[corda-node-4.1-RC03.jar:?]
        at com.r3.corda.finance.obligation.client.flows.UpdateSettlementMethod$Initiator.call(UpdateSettlementMethod.kt:73) ~[?:?]
        at com.r3.corda.finance.obligation.client.flows.UpdateSettlementMethod$Initiator.call(UpdateSettlementMethod.kt:22) ~[?:?]
        at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:269) ~[corda-node-4.1-RC03.jar:?]
        at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:45) ~[corda-node-4.1-RC03.jar:?]
        at co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1092) ~[quasar-core-0.7.10-jdk8.jar:0.7.10]
        at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:788) ~[quasar-core-0.7.10-jdk8.jar:0.7.10]
        at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:100) ~[quasar-core-0.7.10-jdk8.jar:0.7.10]
        at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:91) ~[quasar-core-0.7.10-jdk8.jar:0.7.10]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_212]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_212]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_212]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_212]
        at net.corda.node.utilities.AffinityExecutor$ServiceAffinityExecutor$1$thread$1.run(AffinityExecutor.kt:63) ~[corda-node-4.1-RC03.jar:?]
Caused by: java.lang.ClassNotFoundException: com.r3.corda.finance.obligation.client.flows.MakeOffLedgerPayment
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[?:1.8.0_212]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424) ~[?:1.8.0_212]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_212]
        at net.corda.core.serialization.internal.AttachmentsClassLoader.loadClass(AttachmentsClassLoader.kt:288) ~[corda-core-4.1-RC03.jar:?]
        ... 111 more
```

Reply:
Ah yes, my bad. I forgot about this... To enable CorDapp dependencies to work, the transaction class loader searches for jars containing the needed classes but it will only search JARs with contract implementations in. Presumably we'll get to fixing this in the next point release. Cheers